tezedge_state/src/proposer/mod.rs

```diff
@@ -0,0 +1,948 @@
+        std::mem::replace(&mut self.elapsed, Duration::new(0, 0))
+    }
+}
+
+#[derive(Clone)]
+pub struct TezedgeProposerConfig {
+    pub wait_for_events_timeout: Option<Duration>,
+    pub events_limit: usize,
+    pub record: bool,
+    pub replay: bool,
+}
+
+/// Tezedge proposer wihtout `events`.
+#[derive(Clone)]
+struct TezedgeProposerInner<Efs, M> {
+    config: TezedgeProposerConfig,
+    requests: Vec<TezedgeRequest>,
+    notifications: Vec<Notification>,
+    effects: Efs,
+    time: InternalClock,
+    proposal_persister: Option<ProposalPersisterHandle>,
+    /// Proposal loader for replay.
+    proposal_loader: Option<ProposalLoader>,
+    state: TezedgeStateWrapper,
+    manager: M,
+}
+
+impl<S, NetE, Es, Efs, M> TezedgeProposerInner<Efs, M>
+where
+    S: Read + Write,
+    NetE: NetworkEvent + Debug,
+    Efs: Effects + Debug,
+    M: Manager<Stream = S, NetworkEvent = NetE, Events = Es>,
+{
+    #[inline]
+    fn wait_for_events(&mut self, events: &mut Es) {
+        self.manager
+            .wait_for_events(events, self.config.wait_for_events_timeout)
+    }
+
+    // fn handle_event(&mut self, event: Event<NetE>) {
+    //     match event {
+    //         Event::Tick(at) => {
+    //             accept_proposal!(
+    //                 self.state,
+    //                 TickProposal {
+    //                     time_passed: self.update_time(at),
+    //                     effects: &mut self.effects,
+    //                 },
+    //                 self.proposal_persister
+    //             );
+    //         }
+    //         Event::Network(event) => {
+    //             self.handle_network_event(&event);
+    //         }
+    //     }
+    // }
+
+    fn handle_event_ref<'a>(&mut self, event: EventRef<'a, NetE>) {
+        match event {
+            Event::Tick(at) => {
+                accept_proposal!(
+                    self.state,
+                    TickProposal {
+                        time_passed: self.time.update(at).take_elapsed(),
+                        effects: &mut self.effects
+                    },
+                    self.proposal_persister
+                );
+            }
+            Event::Network(event) => {
+                self.handle_network_event(event);
+            }
+        }
+    }
+
+    fn handle_network_event(&mut self, event: &NetE) {
+        self.time.update(event.time());
+        if event.is_server_event() {
+            // we received event for the server (client opened tcp stream to us).
+            loop {
+                // as an optimization, execute requests only after 100
+                // accepted new connections. We need to execute those
+                // requests as they might include command to stop
+                // listening for new connections or disconnect new peer,
+                // if for example they are blacklisted.
+                for _ in 0..100 {
+                    match self.manager.accept_connection(&event) {
```

Comment:
It will return after each accepted connection, and newer reach 100.

Reply:
Why? If there are 100+ peers in the backlog, we will reach 100, and if not and if we return, that's not a problem. That code is just an optimization.

Reply:
It's documented above those lines.

Reply:
@binier @vlad9486 
I was also staring at the `0..100` I added note to TE-652 for that `check for _ in 0..100 {`

I dont think it is a good idea to have hard-coded constants like this, it should be a part of configuration at least.

is it releated to low/high/expectected_peers count?

Reply:
> is it releated to low/high/expectected_peers count?

No it's not. Basically the problem is that `Acceptor` doesn't have a return type, so we don't instantly know that state machine refused that connection. Also there is no **refuse** incoming connection in tcp, instead we have to accept then drop the connection. So 100 in there is the bound how many peers we will accept at a time in bulk, before `execute_requests` so that we get request to disconnect rejected peers.

That solution works, but obviously it's far from pretty or optimal. I guess we could add associated type to the `Acceptor` for return type, not sure if that will work out well.